image/spec/specs-go/v1/version.go

```diff
@@ -0,0 +1,7 @@
+package v1
+
+const (
+	Version      = "v1.2"
+	VersionMajor = 1
+	VersionMinor = 2
+)
```

Comment:
One thing we should check before this ends up in a release anywhere, is whether the spec matches reality, and if we need a v1.3 (e.g.).

I don't recall when `HealthCheck` was added (pre / post 1.2) etc.

Reply:
Healthcheck is already in 1.2

But yeah, looks like it doesn't mention `OnBuild` and `Shell`.

Reply:
Do we need to do quick check; if any of the new properties were not included in 1.2, maybe we should start this implementation with 1.3 ? WDYT

Also slightly wondering now how this version, and the OCI version interact; would we need a Version in our struct that overrides / shadows the OCI version?


Thinking of cases where someone has a `DockerOCIImage` (v1.2), and wants to take only the `OCI` image (i.e., `ociImage := dockerOCIImage.ImageConfig`)

That image should have the OCI image version that was used, whereas the "outer" image should have the Docker Image version?

Reply:
Oh! I think I'm talking nonsense; the image does not contain the version, right? this is only for the Go package!

🤦 🙈 🙊

Reply:
Yes, there's no version in the image, these are just package-level consts (not sure if they're really needed, but didn't see a reason not to have them either 😄)

Reply:
oh! good one; didn't look where it was there; if it's inside `specs-go` in the OCI, then we should probably follow that.

FWIW: healthcheck were added in v1.2; I see the file was added _because of that_;

- https://github.com/moby/moby/commit/9841cfe4c9af0b63e0c05cd9335bba8bc26a2bb3 https://github.com/moby/moby/pull/25750

_HOWEVER_ we did add the new `StartInterval` field after that, so I'm wondering if we should have 2 commits; one with 1.2, and one with 1.3?

Reply:
We could!

Wondering what to do with `OnBuild` and `Shell`. They seem to be in the Image config **forever**, but are not part of any of the image spec... Were they just accidentally omitted or was it intentional?

Reply:
Oh! Good one on those. So Ideally, we _should_.

I was just comparing between versions, and opened a quick PR to straighten some formatting; https://github.com/moby/moby/pull/46370


With that one, I could do a compare between v1.1 and v1.2, and this is the diff;

```patch
diff --git a/image/spec/v1.1.md b/image/spec/v1.2.md
index 0a473da6a4..cce8c843d6 100644
--- a/image/spec/v1.1.md
+++ b/image/spec/v1.2.md
@@ -1,4 +1,4 @@
-# Docker Image Specification v1.1.0
+# Docker Image Specification v1.2.0
 
 An *Image* is an ordered collection of root filesystem changes and the
 corresponding execution parameters for use within a container runtime. This
@@ -6,7 +6,7 @@ specification outlines the format of these filesystem changes and corresponding
 parameters and describes how to create and use them for use with a container
 runtime and execution tool.
 
-This version of the image specification was adopted starting in Docker 1.10.
+This version of the image specification was adopted starting in Docker 1.12.
 
 ## Terminology
 
@@ -326,6 +326,62 @@ whitespace. It has been added to this example for clarity.
                 not specified, then the first entry of the <code>Cmd</code>
                 array should be interpreted as the executable to run.
             </dd>
+            <dt>
+                Healthcheck <code>struct</code>
+            </dt>
+            <dd>
+                A test to perform to determine whether the container is healthy.
+                Here is an example:
+<pre>{
+  "Test": [
+      "CMD-SHELL",
+      "/usr/bin/check-health localhost"
+  ],
+  "Interval": 30000000000,
+  "Timeout": 10000000000,
+  "Retries": 3
+}</pre>
+                The object has the following fields.
+                <dl>
+                    <dt>
+                        Test <code>array of strings</code>
+                    </dt>
+                    <dd>
+                        The test to perform to check that the container is healthy.
+                        The options are:
+                        <ul>
+                            <li><code>[]</code> : inherit healthcheck from base image</li>
+                            <li><code>["NONE"]</code> : disable healthcheck</li>
+                            <li><code>["CMD", arg1, arg2, ...]</code> : exec arguments directly</li>
+                            <li><code>["CMD-SHELL", command]</code> : run command with system's default shell</li>
+                        </ul>
+                        The test command should exit with a status of 0 if the container is healthy,
+                        or with 1 if it is unhealthy.
+                    </dd>
+                    <dt>
+                        Interval <code>integer</code>
+                    </dt>
+                    <dd>
+                        Number of nanoseconds to wait between probe attempts.
+                    </dd>
+                    <dt>
+                        Timeout <code>integer</code>
+                    </dt>
+                    <dd>
+                        Number of nanoseconds to wait before considering the check to have hung.
+                    </dd>
+                    <dt>
+                        Retries <code>integer</code>
+                    <dt>
+                    <dd>
+                        The number of consecutive failures needed to consider a container as unhealthy.
+                    </dd>
+                </dl>
+                In each case, the field can be omitted to indicate that the
+                value should be inherited from the base layer. These values act
+                as defaults and are merged with any specified when creating a
+                container.
+            </dd>
             <dt>
                 Volumes <code>struct</code>
             </dt>
```

Reply:
So, perhaps we should create new versions of the spec for each of them (if they shipped in different versions of docker, and not "at the same time")

I think we can take 2 approaches;

### Separate files

Continue the "old" approach, and create new files for each spec; while this "works", it does make it harder to follow, and adds some clutter. If we _do_ take this approach, we should take this flow;

- create a new file (e.g. `1.3.0.md`) with no changes (perhaps only the version update) (so a straight "copy"), and commit
- make the changes in the spec for the new version (and commit)

That way it's easier to see "what changed"

### Merge the files

Merge the `1.1` and `1.2` files (1.0 is deprecated, so we should keep that one separate; the diff is also too large to make sense probably)

We need to look how to do the merge to keep as much history as possible (possibly that means

- merging `1.2` into `1.1` (mentioning that commit 9841cfe4c9af0b63e0c05cd9335bba8bc26a2bb3 was applied)
- perhaps add a "history" (although I guess we can tag each revision afterwards once it's in its own repository)
- remove the `1.2` file
- rename the `1.1` file to `specs.md` (unversioned)
- add the new fields (one change at a time), and update the version for each

If each of those is a separate commit, we preserve some history, and _possibly_ could tag older versions after-the-fact in the new repository

Reply:
I did a check on history:

- the image spec v1.0.0 was added in https://github.com/moby/moby/pull/9560
- ONBUILD was added in https://github.com/moby/moby/pull/3254
    - This was before image spec v1.0.0, so "overlooked" ?
- ArgsEscaped was added in https://github.com/moby/moby/commit/9db5db1b94bc1000d151315851096dcc6cd3512d / https://github.com/moby/moby/pull/17851
    - First release: v1.10.0-rc1
    - This was after v1.0.0, but before v1.1.0; https://github.com/microsoft/docker/tree/9db5db1b94bc1000d151315851096dcc6cd3512d/image/spec (so v1.1.0 should be amended)
    - Adopted by OCI in https://github.com/opencontainers/image-spec/pull/892
- SHELL was added in https://github.com/moby/moby/commit/b18ae8c9ccc2eb6cf8aa947f25eb6f1d20089776 / https://github.com/moby/moby/pull/22489
    - First release: v1.12.0-rc1
    - This was after v1.1.0, but before v1.2.0; https://github.com/microsoft/docker/tree/b18ae8c9ccc2eb6cf8aa947f25eb6f1d20089776/image/spec (so a "v1.2.1" ?)

So, in order of changes;

- Add Onbuild to v1.0.0, v1.1.0, and v1.2.0 ("bugfix")
- Add ArgsEscaped to v1.0.0 and make it version v1.0.1
- Add ArgsEscaped to v1.1.0 ("bugfix")
- Add ArgsEscaped to v1.2.0 ("bugfix")
- Add Shell to v1.1.0 and make it v1.1.1
- Add Shell to v1.2.0 ("bugfix")